kernel/src/cpu/vc.rs

```diff
@@ -257,3 +257,369 @@ fn vc_decode_insn(ctx: &X86ExceptionContext) -> Result<Instruction, SvsmError> {
 fn vc_decoding_needed(error_code: usize) -> bool {
     !(SVM_EXIT_EXCP_BASE..=SVM_EXIT_LAST_EXCP).contains(&error_code)
 }
+
+#[cfg(test)]
+mod test {
+    use crate::cpu::msr::{rdtsc, rdtscp, RdtscpOut};
+    use crate::cpu::percpu::this_cpu_mut;
+    use crate::sev::ghcb::GHCB;
+    use crate::sev::utils::{get_dr7, raw_vmmcall, set_dr7};
+    use core::arch::asm;
+    use core::arch::x86_64::__cpuid_count;
+
+    #[test]
+    #[cfg_attr(not(test_in_svsm), ignore = "Can only be run inside guest")]
+    fn test_has_memory_encryption_info_cpuid() {
+        const CPUID_EXTENDED_FUNCTION_INFO: u32 = 0x8000_0000;
+        const CPUID_MEMORY_ENCRYPTION_INFO: u32 = 0x8000_001F;
+        let extended_info = unsafe { __cpuid_count(CPUID_EXTENDED_FUNCTION_INFO, 0) };
+        assert!(extended_info.eax >= CPUID_MEMORY_ENCRYPTION_INFO);
+    }
+
+    #[test]
+    #[cfg_attr(not(test_in_svsm), ignore = "Can only be run inside guest")]
+    fn test_has_amd_cpuid() {
+        const CPUID_VENDOR_INFO: u32 = 0;
+
+        let vendor_info = unsafe { __cpuid_count(CPUID_VENDOR_INFO, 0) };
+
+        let vendor_name_bytes = [vendor_info.ebx, vendor_info.edx, vendor_info.ecx]
+            .map(|v| v.to_le_bytes())
+            .concat();
+
+        assert_eq!(core::str::from_utf8(&vendor_name_bytes), Ok("AuthenticAMD"));
+    }
+
+    const GHCB_FILL_TEST_VALUE: u8 = b'1';
+
+    fn fill_ghcb_with_test_data() {
+        let ghcb = this_cpu_mut().ghcb_unsafe();
+        unsafe {
+            // The count param is 1 to only write one ghcb's worth of data
+            core::ptr::write_bytes(ghcb, GHCB_FILL_TEST_VALUE, 1);
+        }
+    }
+    fn verify_ghcb_was_altered() {
+        let ghcb = this_cpu_mut().ghcb_unsafe();
+        let ghcb_bytes: &[u8; core::mem::size_of::<GHCB>()] = unsafe { &*(ghcb as *const _) };
+        let changed_byte = ghcb_bytes.iter().find(|&&v| v != GHCB_FILL_TEST_VALUE);
+        assert!(changed_byte.is_some());
```

Comment:
This is creating an intermediate reference to the GHCB. I'd favor just using `slice::from_raw_parts()`, which is also less convoluted in my opinion. Also, you can test that at least one byte changed more easily with `Iterator::any()`:

```suggestion
        let ghcb_bytes =
            unsafe { core::slice::from_raw_parts(ghcb.cast::<u8>(), core::mem::size_of::<GHCB>()) };
        assert!(ghcb_bytes.iter().any(|v| *v != GHCB_FILL_TEST_VALUE));
```